src/main/java/io/cdap/plugin/gcp/spanner/source/SpannerSource.java

```diff
@@ -241,11 +252,41 @@ private Schema getSchema(FailureCollector collector) {
                                                          projectId)) {
       DatabaseClient databaseClient =
         spanner.getDatabaseClient(DatabaseId.of(projectId, config.instance, config.database));
-      Statement getTableSchemaStatement = SCHEMA_STATEMENT_BUILDER.bind(TABLE_NAME).to(config.table).build();
+      Map<String, String> columnNameMap = new HashMap<>();
+      // get columns from import query when query does not contain the '*' or 'case'
+      if (config.importQuery != null && !config.importQuery.contains("*") &&
+        !config.importQuery.toLowerCase().contains(" case ")) {
+        Matcher extractedColumnName = TABLE_COLUMNS_PATTERN.matcher(config.importQuery);
+        if (extractedColumnName.find() && extractedColumnName.groupCount() == 1) {
+          String selectedColumns = extractedColumnName.group(1);
+          String[] columns = selectedColumns.split(",");
+          for (String column : columns) {
+            // check for column aliases
+            if (column.toLowerCase().contains(" as ")) {
+              String[] columnNameAndAlias = column.split(COLUMN_ALIAS_SPLIT_PATTERN);
+              columnNameMap.put(columnNameAndAlias[0].trim(), columnNameAndAlias[1].trim());
+            } else {
+              columnNameMap.put(column.trim(), column.trim());
+            }
+          }
+        }
+      }
+      Statement getTableSchemaStatement = null;
+      if (columnNameMap.isEmpty()) {
+        getTableSchemaStatement = SCHEMA_STATEMENT_BUILDER.bind(TABLE_NAME).to(config.table).build();
+      } else {
+        getTableSchemaStatement = SCHEMA_STATEMENT_BUILDER.append(SCHEMA_STATEMENT_COLUMN_FILTER)
+          .bind(TABLE_NAME).to(config.table).bind(TABLE_COLUMNS)
+          .toStringArray(columnNameMap.keySet()).build();
```

Comment:
what happens when column map is empty? Has that case been tested?

Reply:
In the line 279 if the column map is empty it will retrieve all the columns.